### ISSUE TITLE
Loses Deck Versions on startup

### ISSUE THREAD

**Issue by shamanix7** [Epix37/Hearthstone-Deck-Tracker#1528]
On startup the deck versions (that are synchronized frequently from Hearthstats) get lost. Only the version 1.0 is known.
My workflow is to create and edit decks on Hearthstats, import them into Deck Tracker. Frequently I snyc back the statistics to Hearthstats.
So if I have new version of the same deck, Hearthstats have problems with it after restart.

Further, since Version 0.11.3 'Sync with Hearthstats' always hangs and I have to kill Deck Tracker (although the content is synced afterwards). 'Force full sync' works well, so normally I use it instead of normal sync.

I'm using Windows 7, starting Deck Tracker in admin mode.

**Comment by Epix37** [comment]
> My workflow is to create and edit decks on Hearthstats, import them into Deck Tracker. Frequently I snyc back the statistics to Hearthstats.
So if I have new version of the same deck, Hearthstats have problems with it after restart.

Could you do this to reproduce the issue and then post the log here? ([see here](https://github.com/Epix37/Hearthstone-Deck-Tracker/wiki/Reporting-Issues))

**Comment by shamanix7** [comment]
Following the login file of the period when the problems were most severe. I had to split it because it was too big for pastebin:
http://pastebin.com/c38UtpTq
http://pastebin.com/2TaC4xJi

**Comment by shamanix7** [comment]
It seems like the log is getting overridden each time I start hearthstat since the 23rd of September. Maybe this is as it should be.
But just now I had the problem that my deck that I imported from hearthstats got lost (and all it's matches). After restart and force full sync, everything was there again.
Unfortunately, I have only the log file after the restart:
http://pastebin.com/vpV5Jasc

**Comment by shamanix7** [comment]
The behaviour of resetting Version 1.1 Decks to the Cards of Version 1.0 occurs each time I restart HDT (quite annoying). After 'force full sync' this wrong version 1.1 is even written to hearthstats so that everything looks like Version 1.0. The only way to recover my 1.1 Decks is through Hearthstone itself (if I have it there).

**Comment by shamanix7** [comment]
It seems like the log is getting overridden each time I start hearthstat since the 23rd of September. Maybe this is as it should be.
But just now I had the problem that my deck that I imported from hearthstats got lost (and all it's matches). After restart and force full sync, everything was there again.
Unfortunately, I have only the log file after the restart:
http://pastebin.com/vpV5Jasc

**Comment by shamanix7** [comment]
Problem still exists with Version 0.13.0
One of my main Deck is getting overriden at every startup of HDT. I sync with hearthstats (full snyc) afterwards to get the deck version 1.1 again (but I get 1.0 and edit it in HDT manually) and play games which are recorded correctly to version 1.1. After a restart of HDT, there is no version 1.1, just the plain first version of the deck and nothing more is known to HDT. At startup, the PlayerDecks.xml is overriden accordingly...

Probably this is coupled with an issue of Hearthstats, where the version 1.0 and 1.1 of this deck is shown both with the cards of deck version1.0.
There is a button 'Newest Version' (visible only when I clicked one of the versions) which is showing me the correct deck of version 1.1

**Comment by Epix37** [comment]
Try this: 

* Create a backup (options > tracker > backups).
* Make sure `HEARTHSTATS > AUTO DELTE DECKS` is disabled and delete the local version of the deck in HDT.
* Sync and see if the issue remains.

If it does:
* Note: __This will delete the deck and all it's games on HearthStats. If you are missing a lot of games locally, don't want to lose them and can live with the way it currently is, don't do this.__ 
* Restore the backup.
* Select the deck and delete it on HearthStats via `HEARTHSTATS > DELETE SELECTED DECK`.
* Sync the local version to hearthstats again.

Another way of doing this would be to create a new deck with the same cards and then to move the games to that deck (via STATS > GAMES > Deck > Games > move to other deck). __This will move the games on HearthStats as well__ (i.e. remove them from the original deck).

**Comment by Bidouleroux** [comment]
I had this issue after I uploaded changes to a deck from another computer using the same version of HDT (v0.11.3) and it persisted until today on that particular deck. Creating a backup and restoring it as you suggested didn't work, so I restarted and did a force full sync to get the latest version of the deck, which I then deleted on Hearthstats. After that I immediately re-synced to re-upload the deck, then restarted HDT. The deck wasn't there after I restarted, but I re-synced again and was asked to download a new deck. The deck and all its versions were there, but all stats were missing so it seems they weren't uploaded correctly. The deck doesn't disappear anymore after I restart, so there's that.

Dunno if it might be related, but there are also 18 games that can't seem to sync at all. The logs show the same errors each time I sync: http://pastebin.com/YiySrDN4

P.S. Okay, I just try checking the deck on Hearthstats and the website throws an error. All my other decks are fine.

**Comment by chriskarel** [comment]
So, I've been having this exact same issue for months, with a variety of decks.  Sometimes I edit a deck on the HearthStats website.  Sometimes I edit it from HDT directly.  And I could do that from either of two different machines.  End result, is that the decks are all out of wack.  Every time I start HDT, The deck versions have reverted.  Doing a full sync restores the latest versions saved on HearthStats.

Checking through the XML files, it appears that HDT is not properly updating PlayerDecks.xml.  It will update the modify timestamp on HDT start, but not actually make any changes.  When a full sync is done, the updated decks are stored in memory, but never written to disk.  Hence on restart, the newer decks are lost.  File permissions for PlayerDecks.xml seem sane.

I simply deleted the PlayerDecks.xml file, re-opened HDT, then did a HearthStats Full Resync.  This wrote a brand new PlayerDecks.xml file, with all the latest deck info.  Restarting HDT afterwards then had all that latest info.

**Comment by Epix37** [comment]
While that sounds plausible (and I had hope for a second!), PlayerDecks.xml is definitely saved after new versions or changes are downloaded.

**Comment by shamanix7** [comment]
I tried your suggestion:
> Another way of doing this would be to create a new deck with the same cards and then to move the  > games to that deck (via STATS > GAMES > Deck > Games > move to other deck). This will move the > games on HearthStats as well (i.e. remove them from the original deck). 

But I deleted the deck on Hearthstats, assuming the newly created deck on HDT with all the moved games get synchronized with Hearthstats again. But that didn't happen. On Hearthstats I have zero game for the new deck. That's quite annoying. 
Is there a way to get the games synced to Hearthstats?

**Comment by chriskarel** [comment]
Epix, I can pretty reliably reproduce this.  For me, doing a Full Sync only updates the config.xml and DeckStats.xml file.  (based on modify timestamp AND contents)

Starting HDT will also update the timestamps for PlayerDecks.xml and plugins.xml, as well as the above 2. But since it's never getting written after that full sync, it always starts with old versions.

I'm having this problem on both a Win7 and Win8 machine.  If you can't reproduce, is there a debug mode I can flip on to give you more data?  I've clearly gotten my install into some sort of weird state.

**Comment by leadingmore** [comment]
I am having this issue consistently when I try to sync a new deck between two computers.  Yesterday I did a little research into the log file and here is what I found:

When I try running force full sync to get the new version of the deck to download, it updates the deck window, however, in the log, it shows some XML error while trying to write the changes to files.

As a result, on HDT startup, the old version was loaded.

Here is the part of the log that's relevant:

10:19:04 PM|Info|HearthStatsAPI.GetDecksAsync >> getting decks since 0
10:19:15 PM|Info|HearthStatsManager.SyncAsync >> Checking for new versions...
10:19:15 PM|Info|HearthStatsManager.SyncAsync >> saved 1 new versions (1.4.0.0) to Reno Combolock (Warlock)
10:19:16 PM|Error|XmlManager.Save >> Error saving file: C:\Users\Dingmou Li\AppData\Roaming\HearthstoneDeckTracker\PlayerDecks.xml
System.InvalidOperationException: There was an error generating the XML document. ---> System.InvalidOperationException: A circular reference was detected while serializing an object of type Hearthstone_Deck_Tracker.Hearthstone.Deck.
   at System.Xml.Serialization.XmlSerializationWriter.WriteStartElement(String name, String ns, Object o, Boolean writePrefixed, XmlSerializerNamespaces xmlns)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriterDeckList.Write8_Deck(String n, String ns, Deck o, Boolean isNullable, Boolean needType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriterDeckList.Write8_Deck(String n, String ns, Deck o, Boolean isNullable, Boolean needType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriterDeckList.Write10_DeckList(String n, String ns, DeckList o, Boolean isNullable, Boolean needType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriterDeckList.Write11_Decks(Object o)
   --- End of inner exception stack trace ---
   at System.Xml.Serialization.XmlSerializer.Serialize(XmlWriter xmlWriter, Object o, XmlSerializerNamespaces namespaces, String encodingStyle, String id)
   at System.Xml.Serialization.XmlSerializer.Serialize(TextWriter textWriter, Object o, XmlSerializerNamespaces namespaces)
   at Hearthstone_Deck_Tracker.XmlManager`1.Save(String path, Object obj)
10:19:16 PM|Info|HearthStatsManager.SyncAsync >> Checking for edited decks...
10:19:16 PM|Info|HearthStatsManager.SyncAsync >> edited latest version of Reno Combolock (Warlock)
10:19:16 PM|Error|XmlManager.Save >> Error saving file: C:\Users\Dingmou Li\AppData\Roaming\HearthstoneDeckTracker\PlayerDecks.xml
System.InvalidOperationException: There was an error generating the XML document. ---> System.InvalidOperationException: A circular reference was detected while serializing an object of type Hearthstone_Deck_Tracker.Hearthstone.Deck.
   at System.Xml.Serialization.XmlSerializationWriter.WriteStartElement(String name, String ns, Object o, Boolean writePrefixed, XmlSerializerNamespaces xmlns)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriterDeckList.Write8_Deck(String n, String ns, Deck o, Boolean isNullable, Boolean needType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriterDeckList.Write8_Deck(String n, String ns, Deck o, Boolean isNullable, Boolean needType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriterDeckList.Write10_DeckList(String n, String ns, DeckList o, Boolean isNullable, Boolean needType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriterDeckList.Write11_Decks(Object o)
   --- End of inner exception stack trace ---
   at System.Xml.Serialization.XmlSerializer.Serialize(XmlWriter xmlWriter, Object o, XmlSerializerNamespaces namespaces, String encodingStyle, String id)
   at System.Xml.Serialization.XmlSerializer.Serialize(TextWriter textWriter, Object o, XmlSerializerNamespaces namespaces)
   at Hearthstone_Deck_Tracker.XmlManager`1.Save(String path, Object obj)
10:19:16 PM|Info|HearthStatsManager.SyncAsync >> edited 1 decks
[hdt_log.txt](https://github.com/Epix37/Hearthstone-Deck-Tracker/files/199622/hdt_log.txt)

My current work around is:
[1] Close HDT\
[2] Delete DeckStats.xml, DefaultDeckStats.xml, and PlayerDecks.xml.
[3] Start HDT
[4] Do a force full sync
